lib/class-wp-theme-json-gutenberg.php

```diff
@@ -1608,6 +1608,11 @@ protected function get_layout_styles( $block_metadata ) {
 					foreach ( $base_style_rules as $base_style_rule ) {
 						$declarations = array();
 
+						// Skip outputting base styles for flow and constrained layout types if theme doesn't support theme.json.
+						if ( ! wp_theme_has_theme_json() && ( 'default' === $layout_definition['name'] || 'constrained' === $layout_definition['name'] ) ) {
```

Comment:
Can we achieve the same differently, without using `wp_theme_has_theme_json` here? The `WP_Theme_JSON` class is WordPress agnostic: doesn't use "WordPress utilities" (theme supports, doesn't know what a theme is, etc.).

I don't know the details of how the "layout styles" work, so I may be missing tons of context. However, I'd think this kind of logic (whether a theme has or doesn't a `theme.json`) pertains to the public function `gutenberg_get_global_stylesheet` that then makes the proper calls to the utilities.

Reply:
> The WP_Theme_JSON class is WordPress agnostic

Oooh great to know, I hadn't realised that! Was there any thought of using the class outside of WP, or is it just a separation of concerns thing?

I guess if we want the logic to reside in `gutenberg_get_global_stylesheet` we can pass a flag down through `get_stylesheet`.

Reply:
Hmm, [we're using `current_theme_supports`](https://github.com/WordPress/gutenberg/blob/fix/classic-theme-layouts/lib/class-wp-theme-json-gutenberg.php#L1470) further up in `get_layout_styles`; should the same logic apply there?

Reply:
> separation of concerns thing?

This. It simplifies a lot of things (testing, reasoning about the code, etc.).

> Hmm, [we're using current_theme_supports](https://github.com/WordPress/gutenberg/blob/fix/classic-theme-layouts/lib/class-wp-theme-json-gutenberg.php#L1470) further up in get_layout_styles; should the same logic apply there?

In my view, we shouldn't have done that there.

Reply:
Just wondering how we might implement it differently. Would it be something like moving the logic to the resolver around here? https://github.com/WordPress/gutenberg/blob/2477a6d9d47180ed83e7c5fc1c5e500fdf88f53e/lib/class-wp-theme-json-resolver-gutenberg.php#L297

If so, then it looks like the current approach is for keys to be set within the `$theme_support_data` array, so would we have something like `$theme_support_data['settings']['layout']['disableLayoutStyles']` as a supported `theme.json` key?

Then we might also have a key to flag skipping the base layout styles for default and constrained layouts... but how to do that without getting stuck with awkward naming, or an awkward API shape of the `layout` settings object 🤔

Also, this bug fix is targeting 6.5.3 so we might not want to try to change too much in one go?

Reply:
Is it something that could be returned from [gutenberg_get_layout_definitions()](https://github.com/WordPress/gutenberg/blob/13fb10d03414f7eccd250cc1c71ec0d9b14e8250/lib/class-wp-theme-json-gutenberg.php#L1485-L1485)?

So check for `wp_theme_has_theme_json()` in gutenberg_get_layout_definitions() and just remove them before returning? Or something? 😄

Reply:
Thanks for the ideas folks!

> but how to do that without getting stuck with awkward naming

This is the eternal problem 😂 

Would something like `disableContentWidthStyles` work? It's not technically a theme support, but all classic themes that don't have a theme.json in practice _don't_ support it. We haven't needed to check for the presence of a theme.json file in any other part of this class so far, so perhaps there isn't a need for a more generic `disableThemeJson` type thing.

> Is it something that could be returned from [gutenberg_get_layout_definitions()](https://github.com/WordPress/gutenberg/blob/13fb10d03414f7eccd250cc1c71ec0d9b14e8250/lib/class-wp-theme-json-gutenberg.php#L1485-L1485)?

Hmm that might be messier because we don't want to output base flow/constrained styles but we still want to output gap styles _if_ the theme supports `appearance-tools`.

Reply:
> Would something like disableContentWidthStyles work? 

Feels pretty consistent with `disableLayoutStyles`, so I don't mind it!

Reply:
I just thought of another possibility: we could leverage the [`base-layout-styles` type](https://github.com/WordPress/gutenberg/blob/fix/classic-theme-layouts/lib/global-styles-and-settings.php#L34) because that only gets set when there's no theme.json. The name might be a bit confusing because we also have `baseStyles` in the layout definitions, but it's an existing property so would be a small change.

Reply:
Oh, clever, I like it!